model-optimizer/extensions/back/MatMulNormalizer.py

```diff
@@ -142,3 +145,75 @@ def replace_pattern(graph: Graph, match: dict):
             src = port.get_source()
             port.get_connection().set_source(transpose_copy.out_port(0))
             src.connect(start_port)
+
+
+class SmartReshape_HC_Reshape_MatMul(BackReplacementPattern):
+    """
+    Relaxes hard-coded input of Reshape in such sub-graphs:
+
+    input_1     Constant
+        \       /
+        Reshape    input_2
+           \       /
+              MatMul
+                |
+    """
+    enabled = True
+    force_clean_up = True
+
+    def run_after(self):
+        return [MatMulConstTransposesExtraction]
+
+    def pattern(self):
+        return dict(
+            nodes=[
+                ('output_shape', dict(type='Const')),
+                ('output_shape_d', dict()),
+                ('reshape', dict(type='Reshape')),
+                ('reshape_d', dict()),
+                ('other_input', dict(type=lambda t: t not in ['Reshape', 'Transpose'])),
+                ('other_input_d', dict()),
+                ('matmul', dict(type='MatMul')),
+            ],
+            edges=[
+                ('output_shape', 'output_shape_d'),
+                ('output_shape_d', 'reshape', {'in': 1}),
+                ('reshape', 'reshape_d'),
+                ('reshape_d', 'matmul'),
+                ('other_input', 'other_input_d'),
+                ('other_input_d', 'matmul'),
+            ]
+        )
+
+    def replace_pattern(self, graph: Graph, match: dict):
+        matmul = match['matmul']
+        reshape = match['reshape']
+        other_input_port_idx = 0 if match['matmul'].in_port(0).get_source().node.id == match['other_input'].id else 1
+        shape_source = match['matmul'].in_port(other_input_port_idx).get_source()
+        initial_reshape_pattern = reshape.in_port(1).data.get_value()
```

Comment:
If the reshape target shape is already a result of `Shape` operation we still perform the transformation. And actually modify probably reshape-able graph. Are we ok with that? Am I right that there could be a case when the result of `Shape` is still used incorrectly and the MatMul is still not reshape-able?

Reply:
It is not possible -- we match reshape output pattern as a constant.